lib/api/confighandler.go

```diff
@@ -0,0 +1,336 @@
+		c.adjustOptions(w, r, c.cfg.Options())
+	})
+}
+
+func (c *configMuxBuilder) registerLDAP(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.LDAP())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustLDAP(w, r, config.LDAPConfiguration{})
+	})
+	c.PATCH(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustLDAP(w, r, c.cfg.LDAP())
+	})
+}
+
+func (c *configMuxBuilder) registerGUI(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.GUI())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustGUI(w, r, config.GUIConfiguration{})
+	})
+	c.PATCH(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustGUI(w, r, c.cfg.GUI())
+	})
+}
+
+func (c *configMuxBuilder) adjustConfig(w http.ResponseWriter, r *http.Request) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	cfg, err := config.ReadJSON(r.Body, c.id)
+	r.Body.Close()
+	if err != nil {
+		l.Warnln("Decoding posted config:", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	if cfg.GUI.Password, err = checkGUIPassword(c.cfg.GUI(), cfg.GUI); err != nil {
+		l.Warnln("bcrypting password:", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	waiter, err := c.cfg.Replace(cfg)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	c.finish(w, waiter)
+}
+
+func (c *configMuxBuilder) adjustFolder(w http.ResponseWriter, r *http.Request, folder config.FolderConfiguration) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	var waiter config.Waiter
+	err := unmarshalTo(r.Body, &folder)
+	if err == nil {
+		waiter, err = c.cfg.SetFolder(folder)
+	}
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	c.finish(w, waiter)
+}
+
+func (c *configMuxBuilder) adjustDevice(w http.ResponseWriter, r *http.Request, device config.DeviceConfiguration) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	var waiter config.Waiter
+	err := unmarshalTo(r.Body, &device)
+	if err == nil {
+		waiter, err = c.cfg.SetDevice(device)
+	}
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	c.finish(w, waiter)
+}
+
+func (c *configMuxBuilder) adjustOptions(w http.ResponseWriter, r *http.Request, opts config.OptionsConfiguration) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	var waiter config.Waiter
+	err := unmarshalTo(r.Body, &opts)
+	if err == nil {
+		waiter, err = c.cfg.SetOptions(opts)
+	}
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	c.finish(w, waiter)
+}
+
+func (c *configMuxBuilder) adjustGUI(w http.ResponseWriter, r *http.Request, newGUI config.GUIConfiguration) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	oldGUI := c.cfg.GUI()
```

Comment:
I think the thing with newGUI/oldGUI here is confusing given that newGUI is in fact oldGUI and we're grabbing a *new* oldGUI just to have a cache of the password? Better to just grab the password before unmarshal and pass that explicitly to checkGUIPassword (who doesn't really need to full config structs)?

Reply:
This tiny thing remains and I still think it matters. :) (Just in case you missed it since everything else looks fine)

Reply:
Then I don't get it. I now pass passwords directly to `checkGUIPasswords` and query the wrapper directly for password in config right now (active, old) and pass the unmarshalled password as the new password.

Reply:
Oh I was for some reason looking at an old version, but I have a suggestion for the new version as well, hang on